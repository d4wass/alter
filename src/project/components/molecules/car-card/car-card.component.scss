@use '../../../../base';

.car-card {
  display: flex;
  flex-direction: column;
  align-items: center;
  border-radius: base.$main-border-rad;
  height: 290px;
  width: 245px;
  box-shadow: 0px 4px 4px rgba(0, 0, 0, 0.25);
  margin: 20px;

  img {
    border-radius: 10px 10px 0 0;
    width: 100%;
  }
}
