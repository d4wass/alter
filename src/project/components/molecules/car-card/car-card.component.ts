import { Component, Input } from '@angular/core';

interface CarCard {
  title: string;
  img: string;
}

@Component({
  selector: 'app-car-card',
  template: `
    <div class="car-card">
      <img class="car-card-img" src="{{ carCard.img }}" alt="" />
      <div class="car-card-title">
        <h3>{{ carCard.title }}</h3>
      </div>
    </div>
  `,
  styleUrls: ['./car-card.component.scss']
})
export class CarCardComponent {
  @Input() carCard!: CarCard;
}
