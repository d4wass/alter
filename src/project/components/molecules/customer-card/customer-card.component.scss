@use '../../../../base';

.wrapper-card {
    display: flex;
    flex-direction: column;
    border-radius: base.$main-border-rad;
    box-shadow: 1px 1px 6px rgba(0, 0, 0, 0.1);
    // max-width: 250px;
    // height: 200px;
    padding: 20px;
    margin: 10px;
}

h4, p{
    margin: 0;
}
.header-card {
    display: flex;
    margin-bottom: 50px;
    .user-photo {
        border-radius: 50%;
        margin-right: 40px;
    }
}

.rate-card {
    padding-bottom: 10px;
}