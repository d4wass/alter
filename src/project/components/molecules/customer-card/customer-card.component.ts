import { Component, Input, OnInit } from '@angular/core';

export interface CustomerReview {
  img: string;
  name: string;
  surname: string;
  status: string;
  stats: { trips: number; joined: string };
  review: string;
  rate: number;
}

@Component({
  selector: 'app-review-card',
  template: `
    <div class="wrapper-card">
      <div class="content-card">
        <div class="header-card">
          <img src="assets/reviewUser.png" alt="" class="user-photo" />
          <div class="customer-info">
            <h4 class="customer-name">{{ review.name }} {{ review.surname }}</h4>
            <p class="customer-status">{{ review.status }}</p>
            <p class="customer-stats">
              {{ review.stats.trips }} trips &bull; Joined {{ review.stats.joined }}
            </p>
          </div>
        </div>
        <!-- separate component -->
        <div class="rate-card">
          <img src="assets/star-solid 4.svg" *ngFor="let item of settedRate" />
          <img src="assets/star-regular 1.svg" *ngFor="let item of rate" />
        </div>
        <!-- separate component -->
        <p class="review-card">{{ review.review }}</p>
      </div>
    </div>
  `,
  styleUrls: ['./customer-card.component.scss']
})
export class ReviewCardComponent implements OnInit {
  @Input() review!: CustomerReview;
  rate!: number[];
  settedRate!: number[];

  ngOnInit(): void {
    this.rateRender(this.review.rate);
  }

  private rateRender(review: number): void {
    this.rate = Array(5)
      .fill(0)
      .map((x, i) => i);

    if (review) {
      this.settedRate = Array(review)
        .fill(0)
        .map((x, i) => i);
      this.rate = Array(5 - review)
        .fill(0)
        .map((x, i) => i);
    }
  }
}
