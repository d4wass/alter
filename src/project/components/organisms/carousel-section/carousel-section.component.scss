:host{

}

.carousel-section{
    padding: 50px 0;
    h1{
        margin-bottom: 50px;
    }
}

.carousel-content {
    display: flex;
    justify-content: center;
    align-items: center;
    img {
        height: 50px;
    }

    .car-card-wrapper {
        width: 100%;
        padding: 0 30px;
        display: flex;
        justify-content: space-between;
    }
}