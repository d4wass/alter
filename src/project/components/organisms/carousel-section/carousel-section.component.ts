import { animate, AnimationBuilder, AnimationPlayer, style } from '@angular/animations';
import {
  AfterViewInit,
  ChangeDetectionStrategy,
  ChangeDetectorRef,
  Component,
  ElementRef,
  OnInit,
  QueryList,
  Renderer2,
  ViewChild,
  ViewChildren
} from '@angular/core';

const carCards = [
  { title: 'BMW', img: 'assets/car.png' },
  { title: 'Toyota', img: 'assets/car.png' },
  { title: 'Mazda', img: 'assets/car.png' },
  { title: 'Jeep', img: 'assets/car.png' },
  { title: 'Mercedes-Benz', img: 'assets/car.png' },
  { title: 'Mercedes-Benz', img: 'assets/car.png' },
  { title: 'Mercedes-Benz', img: 'assets/car.png' },
  { title: 'Mercedes-Benz', img: 'assets/car.png' },
  { title: 'Mercedes-Benz', img: 'assets/car.png' },
  { title: 'Mercedes-Benz', img: 'assets/car.png' },
  { title: 'Mercedes-Benz', img: 'assets/car.png' }
];

@Component({
  selector: 'app-carousel-section',
  template: `
    <div class="carousel-section">
      <h1>Browse by make</h1>
      <div class="carousel-content">
        <div (click)="handleClick('-')" class="nav-btn">
          <img src="assets/angle-left.svg" />
        </div>
        <div class="car-card-wrapper" #carousel>
          <app-car-card
            *ngFor="let item of carCards"
            #carouselCards
            [carCard]="item"
          ></app-car-card>
        </div>
        <div (click)="handleClick('')" class="nav-btn">
          <img src="assets/angle-right.svg" />
        </div>
      </div>
    </div>
  `,
  styleUrls: ['./carousel-section.component.scss'],
  changeDetection: ChangeDetectionStrategy.OnPush
})
export class CarouselSectionComponent implements AfterViewInit {
  carCards = carCards;
  time = '250ms ease-in';
  private player!: AnimationPlayer;
  private carouselItems!: HTMLElement[];
  private carouselWidth!: number;
  private counter = 0;
  private offset: number = 0;

  constructor(
    private animationBuilder: AnimationBuilder,
    private cdr: ChangeDetectorRef,
    private renderer: Renderer2
  ) {}
  //todo: whats the difference between viewChild and ViewChildren
  @ViewChild('carousel') private carousel!: ElementRef;
  @ViewChildren('carouselCards') private carouselCards!: QueryList<ElementRef>;

  ngAfterViewInit(): void {
    this.carouselItems = [...this.carousel.nativeElement.children];
    setTimeout(() => (this.carouselWidth = this.carousel.nativeElement.clientWidth));
  }

  //add proper types for all
  handleClick(direction: string): void {
    const itemWidth = this.carouselItems[0].clientWidth;
    const maxOffset = this.carouselItems.length * itemWidth - this.carouselWidth;
    this.offset = this.checkCurrentOffset(direction, itemWidth, maxOffset);
    const carouselAnimation = this.buildAnimation();

    this.carouselItems.forEach((item) => {
      this.player = carouselAnimation.create(item);
      this.player.onDone(() => {});
      this.player.play();
    });
  }

  private buildAnimation() {
    return this.animationBuilder.build([
      animate(this.time, style({ transform: `translateX(${this.offset}px)` }))
    ]);
  }

  private counterDirectionControl(direction: string, maxCounterValue: number): void {
    let counter = this.counter;
    if (direction === '' && this.counter !== 0) {
      --counter;
      this.counter = counter;
    }

    if (direction === '-' && this.counter === maxCounterValue) {
      this.counter = counter;
    } else if (direction === '-') {
      ++counter;
      this.counter = counter;
    }
  }

  private checkCurrentOffset(direction: string, itemWidth: number, maxOffset: number): number {
    let newOffset = this.offset;
    const num = this.carouselWidth / this.carouselItems[0].clientWidth;
    const roundedNum = this.roundNumber(num); // how many entire cards I can display for user
    const numberOfCardsVisibleOnCarousel = this.carouselItems.length - Math.floor(roundedNum) - 1;
    const numberOfNonVisibleCardsOnCarousel =
      this.carouselItems.length - numberOfCardsVisibleOnCarousel;
    const movePxOfLastCard = this.checkIfPartOfItemIsVisible(
      numberOfCardsVisibleOnCarousel,
      itemWidth
    );
    this.counterDirectionControl(direction, numberOfNonVisibleCardsOnCarousel);
    console.log(
      'counter',
      this.counter,
      'number of visible cards',
      numberOfCardsVisibleOnCarousel,
      'how many px move last card',
      movePxOfLastCard,
      'non visible cards on carousel',
      numberOfNonVisibleCardsOnCarousel
    );

    //! rework this conditions
    if (direction === '' && newOffset === 0) {
      console.log('1 offset');
      console.log('maxOffset', maxOffset, 'current Offset', this.offset);

      return (newOffset = 0);
    }
    if (this.counter === numberOfNonVisibleCardsOnCarousel) {
      console.log('3 offset');
      console.log('maxOffset', maxOffset, 'current Offset', this.offset);

      return (newOffset = newOffset - movePxOfLastCard);
    }
    if (direction === '-' && newOffset <= 0) {
      console.log('2 offset');
      console.log('maxOffset', maxOffset, 'current Offset', this.offset);

      return (newOffset = newOffset - itemWidth);
    }
    if (direction === '-' && newOffset <= -maxOffset) {
      console.log('4 offset set offset to max valuxe');
      console.log('maxOffset', maxOffset, 'current Offset', this.offset);

      return (newOffset = -maxOffset);
    }
    if (direction === '' && newOffset > 0) {
      console.log('5 offset');
      console.log('maxOffset', maxOffset, 'current Offset', this.offset);

      return (newOffset = 0);
    }

    return direction ? newOffset - itemWidth : newOffset + itemWidth;
  }

  private roundNumber(num: number): number {
    return Math.round((num + Number.EPSILON) * 100) / 100;
  }

  private checkIfPartOfItemIsVisible(numberOfVisibleCards: number, itemWidth: number): number {
    // after implement should return number
    console.log(this.carouselWidth);
    const visibleItemPartWidth = this.carouselWidth - numberOfVisibleCards * itemWidth;
    const movePx = itemWidth - visibleItemPartWidth;
    return movePx;
  }
}
