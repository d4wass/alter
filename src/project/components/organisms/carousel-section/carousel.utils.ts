export enum CarouselEnum {
  ReviewCarousel = 'ReviewCarousel',
  CarCarousel = 'CarCarousel'
}
