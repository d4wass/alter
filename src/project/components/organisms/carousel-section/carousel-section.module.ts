import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { CarouselSectionComponent } from './carousel-section.component';
import { CarCardComponent } from '../../molecules/car-card/car-card.component';

@NgModule({
  declarations: [CarouselSectionComponent, CarCardComponent],
  imports: [CommonModule],
  exports: [CarouselSectionComponent]
})
export class CarouselSectionModule {}
