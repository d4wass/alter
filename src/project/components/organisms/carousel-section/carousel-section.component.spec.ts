import { createComponentFactory, Spectator } from '@ngneat/spectator';
import { MockComponent } from 'ng-mocks';
import { CarCardComponent } from '../../molecules/car-card/car-card.component';
import { ReviewCardComponent } from '../../molecules/customer-card/customer-card.component';
import { CarouselSectionComponent } from './carousel-section.component';
import { CarouselEnum } from './carousel.utils';

describe('CarouselSectionComponent', () => {
  let component: CarouselSectionComponent;
  let spectator: Spectator<CarouselSectionComponent>;
  const carouselTitle = 'carousel title';
  const carouselLength = 5;

  const createComponent = createComponentFactory({
    component: CarouselSectionComponent,
    declarations: [MockComponent(CarCardComponent), MockComponent(ReviewCardComponent)]
  });

  beforeEach(() => {
    spectator = createComponent();
    component = spectator.component;

    component.carouselItems = createMockCarouselItemsArray(carouselLength, 'item');
    component.carouselTitle = carouselTitle;
    component.carouselType = CarouselEnum.CarCarousel;
    component.ngOnInit();
    spectator.detectChanges();
  });

  it('should match the snapshot', () => {
    expect(spectator.fixture).toMatchSnapshot();
  });

  it('should create', () => {
    expect(component).toBeTruthy();
  });

  it('should display the carousel title', () => {
    const title = spectator.query('h1')?.textContent;
    expect(title).toEqual(carouselTitle);
  });

  it('should display the correct number of car card elements when carouselType is CarCarousel', () => {
    const carCardElements = spectator.queryAll('app-car-card');
    expect(carCardElements.length).toEqual(carouselLength);
  });

  it('should display the correct number of review card elements when carouselType is ReviewCarousel', () => {
    component.carouselType = CarouselEnum.ReviewCarousel;
    spectator.detectChanges();
    const reviewCardElements = spectator.queryAll('app-review-card');
    expect(reviewCardElements.length).toEqual(carouselLength);
  });
});

function createMockCarouselItemsArray<T>(length: number, value: T): T[] {
  return new Array<T>(length).fill(value);
}
