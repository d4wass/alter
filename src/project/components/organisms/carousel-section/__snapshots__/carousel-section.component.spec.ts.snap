// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CarouselSectionComponent should match the snapshot 1`] = `
<app-carousel-section
  CarouselEnum={[Function Object]}
  carouselItems={[Function Array]}
  carouselTitle={[Function String]}
  carouselType={[Function String]}
  items={[Function Array]}
>
  <div
    class="carousel-section"
  >
    <h1>
    </h1>
    <div
      appcarouselanimation=""
      class="carousel-content"
    >
      <div
        class="nav-btn-prev"
      >
        <img
          src="assets/angle-left.svg"
        />
      </div>
      <div
        class="nav-btn-next"
      >
        <img
          src="assets/angle-right.svg"
        />
      </div>
    </div>
  </div>
</app-carousel-section>
`;
